pallets/foreign-investments/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+		/// The default sell price limit for token swaps which defines the
+		/// lowest acceptable buy price.
+		///
+		/// NOTE: Can be removed once we implement a
+		/// more sophisticated swap price discovery.
+		// TODO(@review): Since we will only support stable coins from the
+		// beginning, a global default value could be feasible or do we want to
+		// have better granularity?
+		#[pallet::constant]
+		type DefaultTokenSwapSellPriceLimit: Get<Self::Balance>;
+
+		/// The default minimum fulfillment amount for token swaps.
+		///
+		/// TODO(@review): Since we will only support stable coins from the
+		/// beginning, a global default value could be feasible or do we want to
+		/// have better granularity?
+		///
+		/// NOTE: Can be removed once we implement a more sophisticated swap
+		/// price discovery.
+		// TODO(@review): Since we will only support stable coins from the
+		// beginning, a global default value could be feasible or do we want to
+		// have better granularity?
+		#[pallet::constant]
+		type DefaultTokenMinFulfillmentAmount: Get<Self::Balance>;
+
+		/// The token swap order identifying type
+		type TokenSwapOrderId: Parameter
+			+ Member
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Ord
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// The type which exposes token swap order functionality such as
+		/// placing and cancelling orders
+		type TokenSwaps: TokenSwaps<
+			Self::AccountId,
+			CurrencyId = Self::CurrencyId,
+			Balance = Self::Balance,
+			OrderId = Self::TokenSwapOrderId,
+		>;
+
+		/// The hook type which acts upon a finalized investment decrement.
+		type ExecutedDecreaseInvestHook: StatusNotificationHook<
+			Id = cfg_types::investments::ForeignInvestmentInfo<
+				Self::AccountId,
+				Self::InvestmentId,
+				(),
+			>,
+			Status = ExecutedForeignDecrease<Self::Balance, Self::CurrencyId>,
+			Error = DispatchError,
+		>;
+
+		/// The hook type which acts upon a finalized redemption collection.
+		type ExecutedCollectRedeemHook: StatusNotificationHook<
+			Id = cfg_types::investments::ForeignInvestmentInfo<
+				Self::AccountId,
+				Self::InvestmentId,
+				(),
+			>,
+			Status = ExecutedForeignCollectRedeem<Self::Balance, Self::CurrencyId>,
+			Error = DispatchError,
+		>;
+
+		/// Type which provides a conversion from one currency amount to another
+		/// currency amount.
+		type CurrencyConverter: cfg_traits::SimpleCurrencyConversion<
+			Balance = Self::Balance,
+			Currency = Self::CurrencyId,
+			Error = DispatchError,
+		>;
+	}
+
+	/// Maps an investor and their `InvestmentId` to the corresponding
+	/// `InvestState`.
+	///
+	/// NOTE: The lifetime of this storage starts with initializing a currency
+	/// swap into the required pool currency and ends upon fully processing the
+	/// investment after the potential swap. In case a swap is not required, the
+	/// investment starts with `InvestState::InvestmentOngoing`.
+	#[pallet::storage]
+	pub(super) type InvestmentState<T: Config> = StorageDoubleMap<
```

Comment:
The usage of this requires in the code a `unwrap_or_default()` later. I think you can add a `ValueQuery` to this storage then